packages/slate/src/interfaces/text.ts

```diff
@@ -0,0 +1,91 @@
+import isPlainObject from 'is-plain-object'
+import { Mark, Path } from '..'
+
+/**
+ * `Text` objects represent the nodes that contain the actual text content of a
+ * Slate document along with any formatting marks. They are always leaf nodes in
+ * the document tree as they cannot contain any children.
+ */
+
+interface Text {
+  text: string
+  marks: Mark[]
```

Comment:
I'm a little concerned by the performance implications of swapping this from `Immutable.Set` to an Array. In particular the cost of comparing adjacent text nodes to see if the marks are equal is now `O(n^2)` or worse, as seen in the `matches` implementation below:
```
for (const m of existing) {
  if (!Mark.exists(m, marks)) {
    return false
  }
}

for (const m of marks) {
  if (!Mark.exists(m, existing)) {
    return false
  }
}
```

I'm not sure what representation `Immutable.Set` uses but I'm pretty sure `set1.equals(set2)` is not quadratic.

Reply:
@TheSpyder That's a good concern. I have a feeling we might be able to make it not matter, because I think in 90% of cases either:

a. The text nodes will both have no marks.
b. One text node will have marks and the other won't.

So if we do a check for mismatching lengths, or both zero lengths we can short-circuit early in the common cases and not invoke the iterators at all?

Reply:
Certainly those two changes will help a lot! Since we can't easily sort the Marks array, the other way to do this without checking both directions is clone the first array then remove every element found in the second. Further reading:

https://stackoverflow.com/q/245509/7546
https://cs.stackexchange.com/q/69064